### DISCUSSION TITLE
setup/teardown functions like `beforeEach` should be isolated to `test`

### DISCUSSION THREAD

**Discussion by Trinovantes** [vitest-dev/vitest#4120] [Ideas]
I think functions like `beforeEach` should be moved under `test` so that `test.extend` makes sense

```ts
testA = test.extend<FixturesA>({ ... })
testB = test.extend<FixturesB>({ ... })

beforeEach(({ /* whose fixtures will be available here? */ }) => {
  // who executes this?
})

testA('test a', () => {})
testB('test b', () => {})
```

I think right now both tests will execute the `beforeEach` but I think it would be nice to have more fine-grained control over who executes that block as well as being able to automatically infer the types of the fixtures

This will make the API consistent with playwright that many people use for e2e testing. It would be nice for all tests in the project to have consistent syntax (playwright only supports `test.beforeEach`).

**Comment by sheremet-va** [comment]
> whose fixtures will be available here? 

Respected test's fixture (so, if a declares "foo", and b declares "bar", then it will be `{ foo }` before a, and `{ bar }` before b). I do agree this is not ideal, and I like the proposed solution 👍🏻

**Comment by kii-dot** [comment]
If not mistaken, this is how jest designs it and its more intuitive as we want to be able to isolate to tests. 

afterEach is important too. (Y) to proposed solution, something along the lines of
```
describe("testA", () => {
beforeEach(() => {
// some functions
})
describe(() => {
 // Tests in this loop runs beforeEach
})
})
```
